README.md

```diff
@@ -16,17 +16,33 @@ Once you have implemented the `sa` program (and tested it to the best of your ab
 
-*Which algorithm did you use to construct the suffix array. What is its running time?*
+The binary-search based exact pattern matching we implemented by two recursive functions (finding the upper and lower bound). These are called in a for loop over the letter indexes in p and return the start and end indexes used for the following letter in p. Since the sa-boundary for letter i+1 in p is nested in the sa-boundary for letter i, this for loop through p's letters narrows (or keeps constant) the boundary range in every loop until finally returning it for the entire pattern.
+It runs in O(m log n + z).
 
 ## Insights you may have had while implementing the algorithm
+Splicing and comparing entire strings is inefficient as compared to comparing letters recursively with the lower and upper bounds from the previous letter in p.
+
+The first search for lower and upper bounds is double work since it checks the entire range of sa.
 
 ## Problems encountered if any
+We used the upper bound as the end of sa rather than one outside, which caused issues for a while.
 
 ## Validation
 
-*How did you validate that the construct and the search algorithm works?*
+To check the correctness of the algorithm we compared the output of our match-algorithm with the output of the linear exact matching algorithm from project one. We used empty strings, uniform ones, random ones and part of existing DNA sequences to test our functions.
 
 ## Running time
+To test the running time of the algorithm, we simulated random DNA sequences with belonging patterns. The running time was measured for building a suffix tree, sorting the tree, getting the suffix array and doing binary-search exact pattern matching. The running time was also measured for the final function consisting of all the steps. The result is shown in the figure below.
+![](figs/time_random.png)
+
+We also tested the running time with single-symbol strings. The result is shown below:
+
+![](figs/time_worst_case.png)
+
+Looking at the scale of the y-axis, we can see, that the running time is around 3 times worse on single-symbol strings for the final algorithm. Sorting the tree and doing the binary-search search doesn't seem to change whether we have random DNA sequences or single-symbol string. Here the running time seems to be linear. We expect to see a running time of O(m log n + z) for binary-search. To see a running time of O(m log n + z) would maybe require to run the experiment with longer sequences than done in this project.
+
+As known the naive approach to build a suffix tree performs very poorly on single-symbol strings and gives us a running time that looks quadratic whereas it performs better on random DNA strings. Here the running time seems linear. 
+Getting the suffix array is also influenced by which strings we investigate. The running time looks quadratic on single-symbol strings and linear on random DNA sequences.
```

Comment:
Why would the search time be quadratic on single-letter strings? You still do $2m$ binary searches. The interval doesn't shrink as rapidly as when the strings are random, but it should still be $O(m\log n)$ and far from $O(n^2)$.